dawn/src/dawn/Optimizer/PassLocalVarType.cpp

```diff
@@ -123,44 +165,64 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
     // global variable
     if(curVarID_.has_value() && expr->isLocal()) {
       int accessedVariableID = iir::getAccessID(expr);
-      // No self-dependencies
-      if(accessedVariableID != *curVarID_) {
-
-        // If the set of dependers for this variable is not there, it means the variable hasn't been
-        // declared
-        DAWN_ASSERT_MSG(dependencyMap_.count(accessedVariableID),
-                        "Variable accessed before being declared.");
-        // Need to register a dependency. Pair to be added is from dependency (variable accessed in
-        // the rhs) to dependent (lhs).
-        dependencyMap_.at(accessedVariableID).insert(*curVarID_);
-
-        // Retrieve the type of the accessed variable
-        const iir::LocalVariableData& data =
-            metadata_.getLocalVariableDataFromAccessID(accessedVariableID);
-        DAWN_ASSERT_MSG(data.isTypeSet(), "Variable accessed before being declared.");
-        iir::LocalVariableType newType = data.getType();
-
-        // Update the type of the current variable with `newType`
-        updateVariableType(*curVarID_, newType, expr->getSourceLocation());
+      recordVariablePair(accessedVariableID, *curVarID_, expr->getSourceLocation());
+    }
+  }
+  void visit(const std::shared_ptr<iir::IfStmt>& ifStmt) override {
+    // If in the conditional expression we are accessing a field, it means that every statement in
+    // the then and else blocks will have a read-dependency on such field.
+
+    for(const auto& readAccess :
```

Comment:
You could consider using structured bindings here
```
for(const auto& [accessID, unused] : ifStmt->getCondStmt()->getData<iir::IIRStmtData>().CallerAccesses->getReadAccesses())
```
but also not very nice to have the `unused` around.

Reply:
I guess there's no elegant way to iterate through the keys only ...

Reply:
yes...